lemmy/nginx/nginx.conf

```diff
@@ -0,0 +1,95 @@
+worker_processes auto;
+
+events {
+    worker_connections 1024;
+}
+
+http {
+    # We construct a string consistent of the "request method" and "http accept header"
+    # and then apply soem ~simply regexp matches to that combination to decide on the
+    # HTTP upstream we should proxy the request to.
+    #
+    # Example strings:
+    #
+    #   "GET:application/activity+json"
+    #   "GET:text/html"
+    #   "POST:application/activity+json"
+    #
+    # You can see some basic match tests in this regex101 matching this configuration
+    # https://regex101.com/r/vwMJNc/1
+    #
+    # Learn more about nginx maps here http://nginx.org/en/docs/http/ngx_http_map_module.html
+    map "$request_method:$http_accept" $proxpass {
+        # If no explicit matches exists below, send traffic to lemmy-ui
+        default "http://lemmy-ui";
+
+        # GET/HEAD requests that accepts ActivityPub or Linked Data JSON should go to lemmy.
+        #
+        # These requests are used by Mastodon and other fediverse instances to look up profile information,
+        # discover site information and so on.
+        "~^(?:GET|HEAD):.*?application\/(?:activity|ld)\+json" "http://lemmy";
+
+        # All non-GET/HEAD requests should go to lemmy
+        #
+        # Rather than calling out POST, PUT, DELETE, PATCH, CONNECT and all the verbs manually
+        # we simply negate the GET|HEAD pattern from above and accept all possibly $http_accept values
+        "~^(?!(GET|HEAD)).*:" "http://lemmy";
+    }
+
+    upstream lemmy {
+        # this needs to map to the lemmy (server) docker service hostname
+        server "lemmy:8536";
+    }
+
+    upstream lemmy-ui {
+        # this needs to map to the lemmy-ui docker service hostname
+        server "lemmy-ui:1234";
+    }
+
+    server {
+        # this is the port inside docker, not the public one yet
+        listen 1236;
+        listen 8536;
+
+        # change if needed, this is facing the public web
+        server_name localhost;
+        server_tokens off;
+
+        gzip on;
+        gzip_types text/css application/javascript image/svg+xml;
+        gzip_vary on;
+
+        # Upload limit, relevant for pictrs
+        client_max_body_size 20M;
+
+        add_header X-Frame-Options SAMEORIGIN;
+        add_header X-Content-Type-Options nosniff;
+        add_header X-XSS-Protection "1; mode=block";
+
+        # Send actual client IP upstream
+        proxy_set_header X-Real-IP $remote_addr;
+        proxy_set_header Host $host;
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
```

Comment:
This part may be broken (or might not) because traefik is proxy in fron if nginx. Traefik already does this and sets these headers, but nginx is probably not allowing/trusting that. So if lemmy does something with the X-Real-IP header, it might be broken.

Reply:
I don't know how to test this.

Reply:
it might not be a problem, but it does look like they might be using it for something because they went through the trouble of adding it. If its broken, then the X-Real-IP won't be showing the ip address of the user, but instead it will show the IP address of traefik (not super useful).

I can think of two ways to test it: 
 * Log the request and header from lemmy somehow.
 * Replace lemmy frontend with a whoami container, whoami always prints the headers, including X-Real-IP and X-Fowarded-For, which if it works, should show your browser's ip address, not traefik ip address.

Reply:
Are such headers viewable in the browser DevTools?

Reply:
they are not, because these are the headers that come from traefik to nginx, your browser is no longer a party to that conversation.

Reply:
I assume lemmy is using this feature to record the real ip address of posts. so can you see it on one of the posts metadata?

Reply:
I tested it by replacing the lemmy-ui container with this:

```
  lemmy-ui:
    image: traefik/whoami
    command: --port 1234
    depends_on:
      - lemmy
    security_opt:
      - no-new-privileges:true
    restart: unless-stopped
```

And then nginx forwards to the whoami, when I load it returns this:

```
Hostname: d8928ea3e027
IP: 127.0.0.1
IP: 172.25.7.3
RemoteAddr: 172.25.7.8:45964
GET / HTTP/1.1
Host: lemmy.maldek.nogn.rymcg.tech
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.5
Connection: close
Dnt: 1
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-Site: none
Sec-Fetch-User: ?1
Upgrade-Insecure-Requests: 1
X-Forwarded-For: 10.13.13.86, 172.25.7.1
X-Forwarded-Host: lemmy.maldek.nogn.rymcg.tech:8443
X-Forwarded-Port: 8443
X-Forwarded-Proto: https
X-Forwarded-Server: docker-vm
X-Real-Ip: 172.25.7.1
```

which you see at the bottom X-Real-Ip, thats the IP address of my docker host where traefik is running (on the host network) so its not returning correctly the users IP address.

Reply:
Actually, X-Forward-For *is* working, although X-Real-Ip is not.

Reply:
I guess I wont worry about for now